core/src/main/java/com/linecorp/armeria/server/RouteBuilder.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.nio.charset.Charset;
+import java.time.Duration;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.logging.ContentPreviewerFactory;
+
+/**
+ * A builder class for binding a {@link Service} fluently. This class can only be instantiated through
+ * {@link ServerBuilder#route()}.
+ *
+ * <p>Call {@link #service(Service)} to build the {@link Service} and return to the {@link ServerBuilder}.
+ *
+ * <pre>{@code
+ * ServerBuilder sb = new ServerBuilder();
+ * sb.route().get("/foo/bar")                              // Configure the first service.
```

Comment:
Instead of `route` returning the builder what do you think about it taking a parameter `Consumer<Route builder>`?

 I think it's similarly small syntax but will work with most code formatting (current style won't work for e.g. google Java format I think) and it is a bit magical that `service` switches back to a `ServerBuilder`.

Reply:
I thought to return the `ServerBuilder` was ok since we are already doing that in the `VirtualHostBuilder`. 
I was actually going to add a method something like `RouteBuilder.apply(Consumer<RouteBuilder>)` later, but you think it's better to add the `Consumer` to `route()` so that the `RouteBuilder` does not return `ServerBuilder`?
And the style that you mentioned is that the `.consumes()` is aligned with `.get()` in the example?

Reply:
For the style, I'm specifically referring to `route` below which lines up with the above `route`. Automatic code formatters would align `route` with `service` above it I think. I think with the suggested `apply`, it would be similar. Having route take a consumer would avoid it since the parameter would be indented and all calls to route would be aligned. 

I don't know if a consumer API has to be the only one but it would be much easier for me since I use automatic Google format for whitespace

Reply:
@anuraaga, thanks for the point!
It was not a good idea that I made the example which violates our code style to just make it look prettier. I will fix the example like:
```java
sb.route()
  .get("/foo/bar")
  .consumes(JSON, PLAIN_TEXT_UTF_8)
  ...
```

I thought about your suggestion, and it was great to have `ServerBuilder.route(Consumer<RouteBuilder>)`. But I thought it again, and just a little bit worried about having `routeBuilder` in the code.
The code might be like:
```java
sb.route(routeBuilder -> {
    routeBuilder.consumes(JSON, PLAIN_TEXT_UTF_8)
                .produces(JSON_UTF_8, PLAIN_TEXT_UTF_8)
                .requestTimeoutMillis(10)
                .....
});
```
So I think it's probably better to keep as it is.
And it's weird that `RouteBuilder.service(service)` returns the `ServerBuilder`  which the `RouteBuilder` instantiated from, as you pointed out, so perhaps, if I change the name to `RouteBuilder.build(service)`, it could be better I think.

The API `RouteBuilder.apply(Consumer<RouteBuilder>)` which I was going to add was to provide convenient APIs for gRPC and Thrift services which handles `Content-type` and `Accept` headers.
Because the `RouteBuilder` wraps a `PathMapping` with the method `withHttpHeadersInfo()`, I thought it would provide useful methods which set the `HttpMethods`, `consumeTypes` and `produceTypes` on behalf of users, but it turned out to be not.
@trustin and I had a discussion about this, and we have decided that we'd better change current `PathMapping`.
A new wrapper (maybe `Route`? in that case I should rename `RouteBuilder` to something else) will have the `PathMapping`, and optionally `HttpMethods`, `consumeTypes` and `produceTypes`. And the wrapper will handle the request by given condition.
so, inside `RouteBuilder`, `PathMapping` is not always wrapped with `withHttpHeadersInfo`, `RouteBuilder.apply(Consumer<RouteBuilder>)` doesn't need anymore. :-)

Reply:
Finally back from Golden Week :)

Ah I didn't even notice I was tricked by the manual code formatting ;) Yeah if it's

```
sb.route()
    .get()
    .service()
    .route()
    .get()
    .service()
```

then it's consistent in most code styles. Personally I feel like the addition of a variable to have indentation per route is a win, but it's up to you. I find this sort of pattern is easy to write, but very hard to read - and usually we'd want to prioritize code readability?

Reply:
I guess breaking the sentence will help readability quite a bit..?

```java
sb.route()
  .get()
  .service();

sb.route()
  .get()
  .service();
```

I still think the version that uses lambda expression makes the code too verbose. It'd be nice if we have a DSL for Kotlin later though.

Reply:
Ah that's a good point, breaking between routes seems like a good option :)

Just for reference, though, I believe `Consumer<RouteBuilder>` results in this kotlin code

```kotlin
sb.route({
  get("/foo")
  consumes(...)
  etc(...)
})
```

so it might still be nice for that DSL

Reply:
Yeah, @minwoox and I talked about adding `ServerBuilder.route(Consumer<RouteBuilder>)` as an overridden method, but we were not sure if it's a good idea to have two `route()` methods with different return types..  Should we have it with different name, like `withRoute(Consumer<RouteBuilder>)`, or would it just complicate the API?

Reply:
Using `with` seems fine to me - it's consistent with `withMutations` in the other PR too :) It'd be good to consolidate on that throughout the codebase, almost always use `withXXX` when accepting a `Consumer`.

Reply:
What do you think @minwoox?

Reply:
Yeah, it sounds a good idea. 😄